rust/origen/src/services/jtag/service.rs

```diff
@@ -2,131 +2,306 @@
+            tclk: {
+                if let Some(grp) = tclk {
+                    grp.to_identifier()
+                } else {
+                    ("tclk".to_string(), 0)
+                }
+            },
+            tdi: {
+                if let Some(grp) = tdi {
+                    grp.to_identifier()
+                } else {
+                    ("tdi".to_string(), 0)
+                }
+            },
+            tdo: {
+                if let Some(grp) = tdo {
+                    grp.to_identifier()
+                } else {
+                    ("tdo".to_string(), 0)
+                }
+            },
+            tms: {
+                if let Some(grp) = tms {
+                    grp.to_identifier()
+                } else {
+                    ("tms".to_string(), 0)
+                }
+            },
+            trstn: {
+                if let Some(grp) = trstn {
+                    grp.to_identifier()
+                } else {
+                    ("trstm".to_string(), 0)
+                }
+            },
+        })
+    }
+
+    pub fn reset(&self, dut: &Dut) -> Result<()> {
+        let tms = PinCollection::from_group(&dut, &self.tms.0, self.tms.1)?;
+        self._reset(&tms)
     }
 
-    pub fn write_ir(&mut self, value: Value) -> Result<()> {
-        let trans = match value {
-            Value::Bits(bits, size) => node!(
-                JTAGWriteIR,
-                match size {
-                    None => bits.len() as u32,
-                    Some(x) => x,
-                },
-                bits.data()?,
-                Some(bits.overlay_enables()),
-                bits.get_overlay()?
-            ),
-            Value::Data(value, size) => node!(JTAGWriteIR, size, value, None, None),
-        };
-        let tid = TEST.push_and_open(trans);
-
-        // This is temporary to create some AST content
-        cycles(5);
-
-        TEST.close(tid)
-            .expect("Closed JTAG write IR trans properly");
+    fn _reset(&self, tms: &PinCollection) -> Result<()> {
+        let n_id = TEST.push_and_open(node!(JTAGReset, self.id, None));
+        self.comment("Resetting JTAG Interface");
+        tms.drive_high().repeat(6);
+        TEST.close(n_id)?;
+        Ok(())
+    }
+
+    // pub fn idle(&self) -> Result<()> {
+    //     // ...
+    // }
+
+    // pub fn move_to_state(&self, state: JTAGState) -> Result<()> {
+    //     // ...
+    // }
+
+    // fn _move_to_state(&self, state: JTAGState, tms: &PinCollection) -> Result<()> {
+    //     // ...
+    // }
+
+    pub fn write_ir(&self, dut: &Dut, transaction: &Transaction) -> Result<()> {
+        let tms = PinCollection::from_group(&dut, &self.tms.0, self.tms.1)?;
+        let tdi = PinCollection::from_group(&dut, &self.tdi.0, self.tdi.1)?;
+        let trans = node!(JTAGWriteIR, self.id, transaction.clone(), None);
+        let n_id = TEST.push_and_open(trans.clone());
+
+        self.comment("Move to Shift-IR");
+        tms.drive_low().cycle();
```

Comment:
Probably can eventually remove this tms low cycle. It's common practice to park the driver in run test/idle.